[Rui Quan] iP

tanruiquan opened the pull request:
# Duke
> "If you spend too much time thinking about a thing, you'll never get it done." -Bruce Lee ([source](https://dansilvestre.com/productivity-quotes/))

Duke helps you to get things done by remembering your tasks so you do not have to. It's,
* Easy to use 😁 
* ~~Simple~~ Intuitive
* _**Super**_ fast

What you need to do is,
1. Download it [here](https://github.com/nus-cs2103-AY2122S1/ip)
2. Run the program
3. Start adding your tasks!

Features:
- [X] Managing tasks
- [X] Managing deadlines
- [ ] Reminders
---
If you are a Java programmer, you can use it to practice your Java too. Here's the `main` method:
```Java
public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

tanruiquan committed:
Add greeting

tanruiquan committed:
Add echo

tanruiquan committed:
Add exit message

tanruiquan committed:
Add List function

tanruiquan committed:
Add add task function

tanruiquan committed:
Update list formatting

tanruiquan committed:
Add mark as done feature

tanruiquan committed:
Add support for 3 types of tasks

tanruiquan committed:
Add error handling

tanruiquan committed:
Add delete feature

tanruiquan committed:
Add save method in Storage class

tanruiquan committed:
Add load method in Storage class

tanruiquan committed:
Refactor Duke class

tanruiquan committed:
Merge master into branch-Level-7

tanruiquan committed:
Fix file not found error

tanruiquan committed:
Add Parser and Command classes to handle date

tanruiquan committed:
Merge branch 'branch-Level-7'

tanruiquan committed:
Merge branch 'branch-Level-8'

tanruiquan committed:
Fix tasks not saving to storage problem

tanruiquan committed:
Add Ui class

tanruiquan committed:
Ensure all I/O is handled by Ui

tanruiquan committed:
Divide classes into packages

tanruiquan committed:
Add Duke class into package

tanruiquan committed:
Add JUnit tests

tanruiquan committed:
Package the App as a JAR file

tanruiquan committed:
Add JavaDoc comments

tanruiquan committed:
Tweak code to comply with coding standard

tanruiquan committed:
Add find function

tanruiquan committed:
Merge branch 'branch-A-CodingStandard'

tanruiquan committed:
Merge branch 'branch-Level-9'

tanruiquan committed:
Tweak code to adhere to coding standards

tanruiquan committed:
Fix Duke unable to run

tanruiquan committed:
Merge branch 'add-gradle-support'

tanruiquan committed:
Update gitignore

tanruiquan committed:
Automate project builds using Gradle

tanruiquan committed:
Merge branch 'branch-A-Gradle'

tanruiquan committed:
Use Checkstyle

tanruiquan committed:
Merge branch 'branch-A-checkstyle'

tanruiquan committed:
GUI: Add dialog boxes

tanruiquan committed:
Add custom behaviour to DialogBox

tanruiquan committed:
Use FXML for GUI

tanruiquan committed:
Merge branch 'branch-Level-10'

tchiong commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,12 @@
+    public static void main(String[] args) {
+        Application.launch(Main.class, args);
+    }
+}
```
Perhaps you could end your files with a new line

tchiong reviewed (commented):
LGTM! The code is well written!

tanruiquan committed:
Update build.gradle to build jar file

tanruiquan committed:
Add unit tests for Storage class

There is no unit tests for Storage class.

Adding unit tests ensures that the behavior of the Storage class is
correct.

Let's add 2 unit tests to test the correctness of the save method
when the save file don't exist and when it does exist.

Manoharan-Ajay-Anand commented on src/main/java/duke/controllers/DialogBox.java:
```diff
@@ -0,0 +1,63 @@
+        setAlignment(Pos.TOP_LEFT);
+    }
+
+    public static DialogBox getUserDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
+
+    public static DialogBox getDukeDialog(String text, Image img) {
+        var db = new DialogBox(text, img);
+        db.flip();
+        return db;
+    }
```
These two public static methods should be documented to explain their uses. I don't really see the benefit of having two static methods to get the different type of dialog boxes and it complicates your architecture unnecessarily. You can have a simple public constructor with an extra boolean parameter such as fromDuke(true if Duke's response, false if user input). Then, the constructor can handle the logic of the orientation

Manoharan-Ajay-Anand replied:
These two public static methods should be documented to explain their uses. I don't really see the benefit of having two static methods to get the different type of dialog boxes and it complicates your architecture unnecessarily. You can have a simple public constructor with an extra boolean parameter such as fromDuke(true if Duke's response, false if user input). Then, the constructor can handle the logic of the orientation

Manoharan-Ajay-Anand commented on src/main/java/duke/controllers/DialogBox.java:
```diff
@@ -0,0 +1,63 @@
+
+        dialog.setText(text);
+        displayPicture.setImage(img);
+    }
```
Please remove the extra empty line on line 38.

Manoharan-Ajay-Anand commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,45 @@
+     *               saving into a file.
+     * @throws DukeException If there is an error.
+     */
+    public abstract String execute(TaskList tasks, Ui ui, Storage storage) throws DukeException;
```
After you have implemented the GUI, none of your Command sub-classes uses the ui parameter. Its generally not a good practice to have a parameter that is completely unused.

Manoharan-Ajay-Anand commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,98 @@
+     * @return True if the tasks are saved and false otherwise.
+     * @throws DukeException If the file can not be found.
+     */
+    public boolean save(TaskList tasks) throws DukeException {
+        Path dirPath = Paths.get("data");
+        boolean dataPathExists = Files.exists(dirPath);
+        if (!dataPathExists) {
+            File dataDir = new File("data");
+            dataDir.mkdir();
+        }
+
+        try {
+            File file = new File(filePath);
+            PrintWriter writer = new PrintWriter(file);
+            tasks.getTasks().forEach(task -> writer.write(task.format() + System.lineSeparator()));
+            writer.close();
+        } catch (FileNotFoundException e) {
+            throw new DukeException("File could not be found");
+        }
+        return true;
+    }
```
Please remove the empty line at line 53.

Manoharan-Ajay-Anand commented on src/main/java/duke/controllers/MainWindow.java:
```diff
@@ -0,0 +1,70 @@
+package duke.controllers;
```
I am not really sure of using 'controllers' in the package name. I know that in JavaFX they call it FXML controller. However, the FXML controller comes under View in MVC as it is in charge of handling events and handling the ui logic.

Manoharan-Ajay-Anand reviewed (commented):
Your code is very neat and it is very easy to follow. Your application architecture is generally simple and easy to follow. Just take note of the following comments. All the best!

luffingluffy commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,111 @@
+     * @return A Command.
+     * @throws DukeException If the user input is not in the correct form.
+     */
+    public static Command parse(String input) throws DukeException {
```
Maybe you can try using a switch statement instead of if-else.

luffingluffy reviewed (commented):
Other than the comments left by the other reviewer, LGTM!

tanruiquan committed:
Use assertions to document important assumptions

Assertions are not present in the codebase.

The use of assertions ensures that the assumptions
about the program is verifiable. An assertion failure
indicates that there is a possible bug as the code has
resulted in a program state that violates an assumption
about how the code should behave.

Let's add assertions into the Command classes to ensure
that the commands are executed properly.

tanruiquan committed:
Improve code quality in the Parser class

The code in the Parser class contains too many magic literal.
The parse method is too lengthy and hard to understand.

Replacing magic literals with constant variable and abstracting
code into seperate methods make the code easier to read and maintain.

Let's replace the input commands in the Parser class with constant
variables and abstract the conditional statement to a method.

This allows us to group similar commands into a single method call for
verification.

tanruiquan committed:
Merge pull request #1 from tanruiquan/branch-A-Assertions

Use assertions to document important assumptions

tanruiquan committed:
Merge branch 'master' into branch-A-CodeQuality

tanruiquan committed:
Merge pull request #2 from tanruiquan/branch-A-CodeQuality

Improve code quality in the Parser class

tanruiquan committed:
Set up CI

tanruiquan committed:
Remove unused Ui class

tanruiquan committed:
Fix DialogBox not showing all tasks problem

tanruiquan committed:
Add mass operation for delete command

tanruiquan committed:
Add mass operation for done command

tanruiquan committed:
Merge pull request #3 from tanruiquan/branch-C-MassOps

Add functionality for mass operations

zzybluebell commented on src/main/java/duke/Duke.java:
```diff
@@ -54,6 +85,107 @@ public void run() {
+
+        //Scroll down to the end every time dialogContainer's height changes.
+        dialogContainer.heightProperty().addListener((observable) -> scrollPane.setVvalue(1.0));
+    }
```
May I check is it a bit long method? :)

tanruiquan committed:
Improve the GUI

tanruiquan committed:
Add a User Guide

tanruiquan committed:
Set theme jekyll-theme-cayman

tanruiquan committed:
Update User Guide

tanruiquan committed:
Merge branch 'master' of https://github.com/tanruiquan/ip

tanruiquan committed:
Update JavaDoc comments

tanruiquan committed:
Refine exception messages

tanruiquan committed:
Refine exception messages

tanruiquan committed:
Edit images
